### ISSUE TITLE
404 page on new note 302 redirect

### ISSUE THREAD

**Issue by DenisValeev** [hackmdio/hackmd#726]
Hi, 

Sometimes I get a 404 page (IIS) when clicking on the `new note` button and this long url opens up.

Current implementation of this function is as follows: 
1. Open `/new` and get `302` with a redirect to a new page with random name
2. Opening this new generated URL from server

You could probably generate UUIDs for pages on the client side without hitting server for redirects. 

My setup currently: 
IIS, reverse-proxy (url-rewrite + ARR) to fetch results from a VM with docker and hackmd on port 3000. Both sides residing on the subfolder called `hackmd`. 

Thanks.

**Comment by SISheogorath** [comment]
Do you have logs of the moment it happens, that you can provide?

This actually sounds more like an async problem/bug than a misconceptionation and actually wouldn't be solved by moving the UUID generation to the client side.

Actually in my eyes it makes no sense to generate the UUID on the clientside. It's unneeded and only adds information to transmit. while we generate the UUID right now in the database model and make sure it's unique, in case of a clientside generation we would need to add code to reject duplicates, verify that it's valid, …

I think the redirect is an elegant way to keep all this out and just needs a fix for the async handling in case of a 404.

**Comment by DenisValeev** [comment]
I don't think it is probable to get a duplicate with a proper UUID generation implementation on the client side. Collision is possible but very improbable. That's the reason you chose to use UUIDs as a security measure to protect notes from curious eyes of strangers. 

I found this piece of advice [here](https://stackoverflow.com/questions/4243959/iis-reverse-proxy-with-rewrites-cant-handle-a-redirect-from-the-server-we-proxy):

> Go to the ARR proxy settings (they're hidden under IIS -> Application Request Routing Cache -> Server Proxy Settings) and uncheck the "Reverse rewrite host in response headers" checkbox. That should be enough.

I will test it tomorrow... But one thing is certain is that sometimes it works (does not throw 404s) and some other times it doesn't which is strange.

**Comment by DenisValeev** [comment]
Turned out that [URLs with a Plus Character Trigger a 404 Error When Using an IIS Reverse Proxy Server](https://confluence.atlassian.com/confkb/urls-with-a-plus-character-trigger-a-404-error-when-using-an-iis-reverse-proxy-server-312740228.html).

Apparently the UUIDs that you generate contain `+` signs as part of base64 transform. I haven't seen any slashes yet, but I guess you removed them because they could mess with web servers as well. There are many url-safe base64 implementations out there. 

My current `web.config` where I added the `allowDoubleEscaping` functionality:
``` xml
<configuration>
    <system.webServer>
        <rewrite>
            <rules>
                <rule name="ReverseProxyInboundRule1" stopProcessing="true">
                    <match url="(.*)" />
                    <action type="Rewrite" url="http://localhost:3000/{R:1}" />
                </rule>
            </rules>
        </rewrite>
        <security>
          <requestFiltering allowDoubleEscaping="true" />
        </security>
    </system.webServer>
</configuration>
```

**Comment by DenisValeev** [comment]
It didn't take long for me to find this `slash` situation: 
https://hackmd-ce.herokuapp.com/MYTgZixgDAjAtAdgCwEMBM9lmSeAOVAE2ngGYBTdbZM/ANgFZ0AjIA==

Look at this:
...dbZ**M/A**NgFZ0AjIA==

But I believe it was properly initialized when I clicked `New`. So, like I said, three things should be safe-replaced: `+`, `/` and `=`.

**Comment by jackycute** [comment]
@DenisValeev Thanks for reporting this.
I found the `/` slash situation from a customer two weeks ago.
And I tried land a fix by changing the encode method using URI charset in `LZString` and it still contains `+`, `-`, `$` which not fix in your case.
https://github.com/pieroxy/lz-string/blob/master/libs/lz-string.js#L15

So, maybe we should consider encode/decode the base64 string with `encodeURIComponent` or safe-replace by ourself as you proposed.

**Comment by SISheogorath** [comment]
We should keep in mind that this will may break existing links to notes if we don't provide backwards compatible code.

**Comment by ccoenen** [comment]
Breaking old links will make everyone unhappy that uses HackMD like a wiki.
I am also against any non-standard encodings of the URL (like "oh well, lets swap / with !). This will make it much harder for other platforms to interface with HackMD.
If LZString is here to stay (I honestly don't understand why we're using LZString of a UUID), then at least it should be `encodeURIComponent`-ed.

**Comment by DenisValeev** [comment]
It's a part of [RFC 4648](https://tools.ietf.org/html/rfc4648#section-5) not some arbitrary safe-replacements. And it's called `base64url`. Uri-encoded base64 defeats the purpose of otherwise beautiful representation of a guid.

> I honestly don't understand why we're using LZString of a UUID

I think it is incorrectly used in this project. A uuid per se is already a pretty short representation, for example, consider `3b0da04e-79c0-4219-a3f4-b8b39b156304`. Properly converted (as binary) to `base64url` it should look like this `Ow2gTnnAQhmj9LizmxVjBA` (shorter and nicer than uuid without hyphens) but we have this:
```
Should be:
3b0da04e-79c0-4219-a3f4-b8b39b156304
3b0da04e79c04219a3f4b8b39b156304
Ow2gTnnAQhmj9LizmxVjBA

Actually happening in hackmd:
lz-compress("3b0da04e-79c0-4219-a3f4-b8b39b156304") = 
02330c30600260860a2c5a60b003801364c6046c46c01255009864cd00444638101515f03680e2600000

base64(0x02330...) = 
AjMMMGACYIYKLFpgsAOAE2TGBGxGwBJVAJhkzQBERjgQFRXwNoDiYAAA
``` 
In conclusion, lz-compression of uuid and conversion to base64 should be removed entirely from this project.

**Comment by jackycute** [comment]
@DenisValeev @ccoenen 
So the `LZString` is used here because this project was originally try to enhance security of collaborate notes.
You could switch to the very early version like 0.4.0, we even use `LZString` in socket.io data transport and db savings, but we found they are not efficient and useful later then removed it.
The note url using it is the only left part which preserve for not breaking any current note links.

As @DenisValeev suggested and I see the RFC now, it's time to solve this once at time.
Will definitely make migration or workaround for this, thank you guys 😄

**Comment by jackycute** [comment]
Let's explain why using base64 in the url.
Don't be too harsh, this is the real thought of a beginner in security two years ago.

Personally, I don't like to expose UUID in the url directly.
I think it gain the attack surface that user can actually see the UUID is id in the db.
UUID or base64 are both common used and easy to decode.
`LZString` is the choice that it compress the content (but longer in base64 format...) and it's harder for user to guess what that is.
And yes, the url length counts. That's why I add `shortid` which provides much shorter id for user to publish it.

**Comment by SISheogorath** [comment]
I agree that exposing the UUID directly is, even when it's from a security perspective irrelevant, ugly.

No matter what we do, we **have to keep backwards compatiblity** for at least 2 years.

We can't break links, that's a total nogo no matter how wrong they are encoded. As @ccoenen already pointed out, people share links, people safe links in favorites, people link between internal or from external documents. So even when we make it to change our linking in future and build an internal migration, it'll not change any of the millions of links that are in Blogs, browsers, wikis, …

There is **no way** to just throw away the current way of link building.

But for the future I'm for something else:
Let's go the long wanted `/user/notename` way to handle URLs in future. We add a new field in the DB, add a name there. This name will be random generated like the shortID right now and can in future be changed into something like filenames. This way the problem is basically solved.

**Comment by ccoenen** [comment]
How about adding a table for redirects? Advantages:
- We don't need a complicated code-path while looking up documents. 404s will simply _also_ check the redirects table (and then issue a `moved permanently` redirect).
- We can also use it in case of renaming or moving a document to a directory
- The note table does not need to be touched
- The user will end up at the canonical url; no two adresses will magically display the same content

**Comment by jackycute** [comment]
I've pushed a PR #742 
Could everyone here help test it?
Thanks 😄

**Comment by DenisValeev** [comment]
@jackycute Thanks for the PR.

```
+      encodeNoteId: function (id) {
+        // remove dashes in UUID and encode in url-safe base64
+        return base64url.encode(id.replace(/-/g, ''))
+      },
```

Shouldn't we use `Buffer` instead of string when encoding? UUIDs are 16 bytes long but compressing as `string` would yield 32 bytes converted to base64, compare:

```
nuIyo7qe80WL0J9aMAqAwQ (as bytes)
OWVlMjMyYTNiYTllZjM0NThiZDA5ZjVhMzAwYTgwYzE (as string)
```
Also, note that bytes in guids are not in byte-to-byte order when converting to Buffer, so it is better to use a version specific uuid parser.

**Comment by jackycute** [comment]
@DenisValeev 
That's an gorgeous reminder! Half length now. (wow such an encoding 👍 )

**Comment by SISheogorath** [comment]
Works well and is part of the new releases :) Thanks for all your work and thoughts!
